### PR TITLE
Bugfix/jdbc type provider initialization

### LINKED ISSUE #246: ResultSet failed to load JsonObject with custom type
### Version

4.2.0

### Context

1. The data loaded into ResultSet from MySql and contain custom type values ```LocalDate```.
In try to ```ResultSet.toJson()``` and then ```new ResultSet(JsonObject)``` got exception.

Following the exception:

```
java.lang.IllegalStateException: Illegal type in Json: class java.time.LocalDate
	at io.vertx.ext.sql.ResultSetConverter.lambda$1(ResultSetConverter.java:52) ~[classes/:?]
	at io.vertx.ext.sql.ResultSetConverter.fromJson(ResultSetConverter.java:50) ~[classes/:?]
	at io.vertx.ext.sql.ResultSet.<init>(ResultSet.java:66) ~[classes/:?]
```

### LINKED ISSUE #250: [regression]JDBCClient with date on error
When using JDBCClient (vs JdbcPool) directly occurred error with date on INSERT (probably UPDATE ?)

Using JdbcPool is ok

`io.vertx.serviceproxy.ServiceException: ORA-01843: ce n'est pas un mois valide`

When debugging, all call with JDBCTypeProvider.fromParameter(xx) return JDBCType.OTHER;
Throwing error : 
```JDBCTypeProvider | ParameterMetadata is unsupported bycom.mchange.v2.c3p0.impl.NewProxyPreparedStatement
	at oracle.jdbc.driver.OracleParameterMetaData.checkValidIndex(OracleParameterMetaData.java:219)
	at oracle.jdbc.driver.OracleParameterMetaData.getParameterType(OracleParameterMetaData.java:330)
	at io.vertx.ext.jdbc.impl.actions.JDBCTypeProvider.lambda$fromParameter$0(JDBCTypeProvider.java:27)
	at io.vertx.ext.jdbc.spi.impl.JDBCEncoderImpl.encode(JDBCEncoderImpl.java:27)
	at io.vertx.ext.jdbc.impl.actions.AbstractJDBCAction.fillStatement(AbstractJDBCAction.java:109)
	at io.vertx.ext.jdbc.impl.actions.JDBCUpdate.execute(JDBCUpdate.java:88)
	at io.vertx.ext.jdbc.impl.actions.JDBCUpdate.execute(JDBCUpdate.java:29)
```
This call is ok on SELECT

```
jdbcClient.getConnection(con -> {
 .... 
  con.updateWithParams(sql, map, res -> {})
})
```

Suppose regression from https://github.com/vert-x3/vertx-jdbc-client/pull/235 ?

### Version

4.2.x (ok with 4.1.5)
oracle driver com.oracle.database.jdbc:ojdbc11-production:21.3.0.0

### Context
INSERT INTO MyTable ( DA ) VALUES ( ? )
[2021-11-05T14:35:13.787791287Z"] => date as Instant.now()
DA column is TIMESTAMP(6)

i'll try to do a reproducer test this week-end

### LINKED ISSUE #255: [regression] JDBCSQLRowStream handles DateTimeOffset type incorrectly
### Version

This bug is in the tree since 4.2.0, 4.1.6 seems ok

### Context

I just tried to update from 4.0.3 to 4.2.1, when streaming a table on a MSSQL server that has a column of type DateTimeOffset, I receive a 

```java
throw new IllegalArgumentException("Type:" + type + " is not a valid Types.java value.");
```

### Do you have a reproducer?

It's quite easy to reproduce, just add a DateTimeOffset column to a MSSQL table and try to stream it using both the legacy or the new  API

### Extra

The code in question is :

4.1.6
```java
private void readBatch() {
    if (!rsClosed.get()) {
      ctx.<List<JsonArray>>executeBlocking(fut -> {
        try {
          List<JsonArray> rows = new ArrayList<>(fetchSize);
          for (int i = 0;i < fetchSize && rs.next();i++) {
            JsonArray result = new JsonArray();
            for (int j = 1; j <= cols; j++) {
              Object res = JDBCStatementHelper.convertSqlValue(rs.getObject(j)); // Seems to work properly
              if (res != null) {
                result.add(res);
              } else {
                result.addNull();
              }
            }
            rows.add(result);
          }
          fut.complete(rows);
        } catch (SQLException e) {
          fut.fail(e);
        }
      }, statementsQueue, ar -> {
        if (ar.succeeded()) {
          List<JsonArray> rows = ar.result();
          if (rows.isEmpty()) {
            empty(null);
          } else if (pending.write(rows)) {
            readBatch();
          }
        } else {
          empty(ar.cause());
        }
      });
    }
```

4.2.1
```java
private void readBatch() {
    if (!rsClosed.get()) {
      ctx.<List<JsonArray>>executeBlocking(fut -> {
        try {
          JDBCTypeProvider provider = col -> JDBCType.valueOf(metaData.getColumnType(col)); // Throws IllegalArgumentException
          List<JsonArray> rows = new ArrayList<>(fetchSize);
          for (int i = 0; i < fetchSize && rs.next(); i++) {
            JsonArray result = new JsonArray();
            for (int j = 1; j <= cols; j++) {
              Object res = decoder.parse(rs, j, provider);
              if (res != null) {
                result.add(res);
              } else {
                result.addNull();
              }
            }
            rows.add(result);
          }
          fut.complete(rows);
        } catch (SQLException e) {
          fut.fail(e);
        }
      }, statementsQueue, ar -> {
        if (ar.succeeded()) {
          List<JsonArray> rows = ar.result();
          if (rows.isEmpty()) {
            empty(null);
          } else if (pending.write(rows)) {
            readBatch();
          }
        } else {
          empty(ar.cause());
        }
      });
    }
  }
```

### PR DESCRIPTION
**Pull request by zero88**
Fix #255 
Fix #250 
Fix #246

### CODE REVIEW

**Comment by zero88** [comment]
Hi @vietj 
I provided a patch and make the code cleaner than the previous version.
Also, I updated the document about `JDBCEncoder` and `JDBCDecoder`. You should look and correct it for a better guide coz my English is not good as a native speaker.

**Comment by zero88** [comment]
Not sure why running Oracle Testcontainers is timeout on GitHub action

File: src/main/java/io/vertx/ext/jdbc/impl/actions/JDBCTypeWrapper.java [hunk 1]
```diff
@@ -0,0 +1,198 @@
+package io.vertx.ext.jdbc.impl.actions;
```
Comment (pmlopes) [inline src/main/java/io/vertx/ext/jdbc/impl/actions/JDBCTypeWrapper.java:1]: For Eclipse Compliance, can you add a copyright header just like the other files?

File: src/main/java/io/vertx/ext/jdbc/spi/JDBCColumnDescriptorProvider.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
+package io.vertx.ext.jdbc.spi;
```
Comment (pmlopes) [inline src/main/java/io/vertx/ext/jdbc/spi/JDBCColumnDescriptorProvider.java:1]: For Eclipse Compliance, can you add a copyright header just like the other files?

File: src/main/java/io/vertx/jdbcclient/impl/actions/JDBCPropertyAccessor.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
+package io.vertx.jdbcclient.impl.actions;
```
Comment (pmlopes) [inline src/main/java/io/vertx/jdbcclient/impl/actions/JDBCPropertyAccessor.java:1]: For Eclipse Compliance, can you add a copyright header just like the other files?

File: src/test/java/io/vertx/it/MSSQLDecoder.java [hunk 4]
```diff
@@ -0,0 +1,30 @@
+package io.vertx.it;
```
Comment (pmlopes) [inline src/test/java/io/vertx/it/MSSQLDecoder.java:1]: For Eclipse Compliance, can you add a copyright header just like the other files?

**Comment by zero88** [comment]
These are some examples

`Decoder`

https://github.com/vert-x3/vertx-jdbc-client/blob/4617b4250448b0c878f6fb28bc9b76a2f81d9785/src/test/java/io/vertx/it/MSSQLDecoder.java#L11-L30

`Encoder`

https://github.com/vert-x3/vertx-jdbc-client/blob/4617b4250448b0c878f6fb28bc9b76a2f81d9785/src/test/java/io/vertx/ext/jdbc/JDBCTypesTestBase.java#L52-L61

`How to register`

https://github.com/vert-x3/vertx-jdbc-client/blob/529a848a5ee6aa0447054b1df85ace8b928e7cd7/src/main/java/examples/JDBCTypeExamples.java#L33-L52

**Comment by pmlopes** [comment]
@zero88 you could try to run the Oracle test behind of a special profile as I am guessing that the containers requirements in terms of cores/ram are a bit too high for github actions.

This profile would then be disabled on CI (for now).

**Comment by zero88** [comment]
@pmlopes these profiles are in `vertx-sql-client`

But I already my test pass with Oracle in local
https://github.com/vert-x3/vertx-jdbc-client/blob/6d186d64fabbd71f492cba051adee8cf8902776a/src/test/java/io/vertx/it/OracleTest.java#L43-L45

**Comment by zero88** [comment]
@pmlopes @vietj could you have a look?

**Comment by pmlopes** [comment]
Thanks! @zero88

**Pull request merged**